### PR TITLE
Support tests for removed relations and units

### PR DESCRIPTION
**Pull request by balbirthomas**
This pull request add support for testing relation departed and relation broken events, by adding methods in the test harness to remove relations and their units.

### CODE REVIEW

File: ops/testing.py [hunk 1]
```diff
@@ -458,6 +485,63 @@ def add_relation_unit(self, relation_id: int, remote_unit_name: str) -> None:
...
+            remote_unit_name: A string representing the remote unit that is being added.
+
+        Raises:
+            KeyError: if relation_id or remote_unit_name is not valid
+            ValueError: if remote_unit_name is not valid
+        """
+        relation_name = self._backend._relation_names[relation_id]
+
+        # gather data to invalidate cache later
+        if self._model is not None:  # defensive check for model
```
Comment (facundobatista) [inline ops/testing.py:517]: But what happens if self._model is None? For sure you can not continue happily with the method, as for example 'unit_cache' will be not defined.

It looks a test is missing for this case, too.

File: ops/testing.py [hunk 2]
```diff
@@ -458,6 +485,63 @@ def add_relation_unit(self, relation_id: int, remote_unit_name: str) -> None:
...
+
+        # statements which could access cache
+        self._emit_relation_departed(relation_id, remote_unit_name)
+        self._backend._relation_data[relation_id].pop(remote_unit_name)
+        self._backend._relation_app_and_units[relation_id][
+            "units"].remove(remote_unit_name)
+        self._backend._relation_list_map[relation_id].remove(remote_unit_name)
+
+        if unit_cache is not None:
+            unit_cache._invalidate()
```
Comment (facundobatista) [inline ops/testing.py:530]: Isn't a better way for doing this than calling a private method?

File: ops/testing.py [hunk 3]
```diff
@@ -458,6 +485,63 @@ def add_relation_unit(self, relation_id: int, remote_unit_name: str) -> None:
...
+
+    def _emit_relation_departed(self, relation_id, unit_name):
```
Comment (facundobatista) [inline ops/testing.py:532]: Please add a docstring here.

**Review by facundobatista — CHANGES_REQUESTED** [review]
Nice! I added a couple of details, thanks!

File: ops/testing.py (hunk 2 above)
Comment (balbirthomas) [reply ops/testing.py:530]: This seems to be the practice adopted elsewhere. See for instance `add_relation_unit`. Is there an alternative you would suggest ?

File: ops/testing.py (hunk 3 above)
Comment (balbirthomas) [reply ops/testing.py:532]: done.

File: ops/testing.py (hunk 1 above)
Comment (balbirthomas) [reply ops/testing.py:517]: I am not sure under what conditions model can be none. Any ideas ? However if model is none then a relation unit will not be added in the first place. See for instance `add_relation_unit`, even though various data structures associated with the relation do get updated (c.f. `add_relation` and `add_relation_unit`). Hence in the process of removing relations the same logic is being inverted. I do not see what else could be done. Further if the `_model` is None then so will `unit_cache` and there is a defensive check for this when invalidating the unit cache.

Comment (balbirthomas) [reply ops/testing.py:517]: After investigating, here is what I find. It is not possible to add or remove a relation by forcing `harness.charm.framework.model` to be `None` because the `restore` method of `RelationEvent` access `model`.

Comment (balbirthomas) [reply ops/testing.py]: Removed check for model being none as per agreement in Standup.

File: ops/testing.py (hunk 2 above)
Comment (jameinel) [reply ops/testing.py:528]: Harness as the way to mutate the Model without actually exposing ways to mutate Model directly was decided as the best trade off to have testability but not runtime mutability.

File: ops/testing.py [hunk 4]
```diff
@@ -405,16 +405,43 @@ def add_relation(self, relation_name: str, remote_app: str) -> int:
...
+
+        Args:
+            relation_name: The relation on Charm that is being related to
+            remote_app: The name of the application that is being related to
+
+        Raises:
+            ValueError: if relation_name is invalid
+            RelationNotFoundError: if valid relation id for relation_name is not found
+        """
+        rel_id = self._backend._relation_ids_map[relation_name][0]
```
[the remark this answers is not in the document]
Comment (balbirthomas) [reply ops/testing.py:419]: @jameinel Good point. Thank you. I agree that allowing developers to remove a relation without bothering to remove each unit would be a more friendly API. I presume this behaviour is also supported by Juju, so behaviour of Harness and Juju are aligned. 

Since this same pull request also implements `remove_relation_unit` would it be acceptable instead to do
```
for unit_name in self._backend._relation_list_map[rel_id]:
    self.remove_relation_unit(rel_id, unit_name)
```

[the remark this answers is not in the document]
Comment (balbirthomas) [reply ops/testing.py:419]: Done. Also modified unit test.

File: ops/testing.py [hunk 5]
```diff
@@ -446,18 +475,74 @@ def add_relation_unit(self, relation_id: int, remote_unit_name: str) -> None:
...
+
+        Example::
+
+          rel_id = harness.add_relation('db', 'postgresql')
+          harness.add_relation_unit(rel_id, 'postgresql/0')
+          ...
+          harness.remove_relation_unit(rel_id, 'postgresql/0')
+
+        This will trigger a `relation_departed` event. This would
+        normally be followed by a `relation_changed` event triggered
```
Comment (jameinel) [inline ops/testing.py:500]: Is this actually true? I don't remember seeing a relation-changed as a result of relation-departed. It is true of relation-joined, but I would think departed means the unit is gone and thus we don't have a changed event related to that unit.

**Review by jameinel — APPROVED** [review]
I'm not sure if the docstring comment is correct. It would surprise me if it is, but that doesn't mean that it isn't actually true.

File: ops/testing.py (hunk 5 above)
Comment (balbirthomas) [reply ops/testing.py:500]: Thank you @jameinel . Ah. Looks like I wrote that backwards. Consulting [ops docs](https://ops.readthedocs.io/en/latest/#ops.charm.RelationDepartedEvent) again, can I rephrase this as saying "This will normally be proceeded by a relation changed event" ...

File: ops/testing.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (javacruft) [reply ops/testing.py:421]: should remove relation not be keyed on relation_id? relations have multiplicity and developers will need to test this type of scenario where individual relations disppear.

File: ops/testing.py [hunk 6]
```diff
@@ -407,16 +407,53 @@ def add_relation(self, relation_name: str, remote_app: str) -> int:
         self._emit_relation_created(relation_name, rel_id, remote_app)
         return rel_id
 
+    def remove_relation(self, relation_name: str, remote_app: str, rel_id=None) -> None:
```
Comment (javacruft) [inline ops/testing.py:410]: rel_id -> relation_id: str=None

makes the parameter name consistent with other methods.

Comment (balbirthomas) [reply ops/testing.py:410]: OK will change parameter name to "relation_id". When writing the charm code isn't the relation name the more developer friendly interface ? If so wouldn't it be appropriate to require a relation id only if the relation name is ambiguous ?

Comment (balbirthomas) [reply ops/testing.py]: Changed parameter name to `relation_id`. Let me know if the above rationale is acceptable. If not I can make `relation_id` a mandatory argument and remove `relation_name` as it will be redundant, since there is a internal map `_TestingModelBackend._relation_names` from relation ids to relation names.

**Comment by jameinel** [comment]
I'm curious why we would have both relation_name and relation_id. If you
consider update_relation we're talking about a very specific relation, so
it only takes a relation_id not a name. I would expect the same for
remove_relation because you are expressly removing the one relation strand
that you added earlier. (not all things related to 'foo')

That is why add_relation returns the relation_id that I would expect you to
pass back into remove_relation. You also don't need the remote_app because
that is already fully defined by relation_id. So just
```
def remove_relation(relation_id: int):
```

**Comment by balbirthomas** [comment]
> I'm curious why we would have both relation_name and relation_id.

Done (relation id is now sole argument). Also added another unit test to check RelationNotFoundError exception is raised if relation id is invalid.

**Review by javacruft — APPROVED** [review]
LGTM - thanks for aligning the semantics of relation removal to my suggestion!

**Comment by balbirthomas** [comment]
> LGTM - thanks for aligning the semantics of relation removal to my suggestion!

It is I who should be thanking your for the very insightful and sharp observations you made. You were spot on. Thank you.

**Review by rbarry82 — APPROVED** [review]
LGTM, thanks!
